Source/Details/ASPINRemoteImageDownloader.m

```diff
@@ -256,51 +252,38 @@ - (nullable id)downloadImageWithURLs:(NSArray <NSURL *> *)URLs
-        /// If we're targeting the main queue and we're on the main thread, complete immediately.
-        if (ASDisplayNodeThreadIsMain() && callbackQueue == dispatch_get_main_queue()) {
-#if PIN_ANIMATED_AVAILABLE
-            if (result.alternativeRepresentation) {
-                completion(result.alternativeRepresentation, result.error, result.UUID);
-            } else {
-                completion(result.image, result.error, result.UUID);
-            }
-#else
-            completion(result.image, result.error, result.UUID);
-#endif
-        } else {
-            dispatch_async(callbackQueue, ^{
+  PINRemoteImageManagerProgressDownload progressDownload = ^(int64_t completedBytes, int64_t totalBytes) {
+    if (downloadProgress == nil) { return; }
+
+    [ASPINRemoteImageDownloader _performWithCallbackQueue:callbackQueue work:^{
+      downloadProgress(completedBytes / (CGFloat)totalBytes);
+    }];
+  };
+
+  PINRemoteImageManagerImageCompletion imageCompletion = ^(PINRemoteImageManagerResult * _Nonnull result) {
+    [ASPINRemoteImageDownloader _performWithCallbackQueue:callbackQueue work:^{
 #if PIN_ANIMATED_AVAILABLE
-                if (result.alternativeRepresentation) {
-                    completion(result.alternativeRepresentation, result.error, result.UUID);
-                } else {
-                    completion(result.image, result.error, result.UUID);
-                }
+      if (result.alternativeRepresentation) {
+        completion(result.alternativeRepresentation, result.error, result.UUID);
+      } else {
+        completion(result.image, result.error, result.UUID);
+      }
 #else
-                completion(result.image, result.error, result.UUID);
+      completion(result.image, result.error, result.UUID);
 #endif
-            });
-        }
-    };
-    
-    return [[self sharedPINRemoteImageManager] downloadImageWithURLs:URLs
-                                                             options:PINRemoteImageManagerDownloadOptionsSkipDecode
-                                                       progressImage:nil
-                                                    progressDownload:progressDownload
-                                                          completion:imageCompletion];
+    }];
+  };
+
+  // add "IgnoreCache" option since we have a caching API so we already checked it, not worth checking again.
+  // PINRemoteImage is responsible for coalescing downloads, and even if it wasn't, the tiny probability of
+  // extra downloads isn't worth the effort of rechecking caches every single time. In order to provide
+  // feedback to the consumer about whether images are cached, we can't simply make the cache a no-op and
+  // check the cache as part of this download.
+  return [[self sharedPINRemoteImageManager] downloadImageWithURLs:URLs
+                                                           options:PINRemoteImageManagerDownloadOptionsSkipDecode | PINRemoteImageManagerDownloadOptionsIgnoreCache
```

Comment:
Sadly I don't think we should skip the cache because the synchronous cache check only checks the memory cache not the disk cache.

Reply:
But we'll check the disk cache in `cachedImageWithURL:callbackQueue:completion:` so I think we're OK?

Reply:
I take this back, didn't see that the cache check above was implemented.

Reply:
For history, @Adlai-Holler and I discussed and we think this is safe. He pointed out the only case where this could cause an extra download is:
```
- request0 -> cache: miss
- request0 -> download: started
- request1 -> cache: miss
- request0 -> download: finished
- request1 -> download: started
```

This seems *very* unlikely in practice.